File: frontend/src/api/problem-details.ts
export type ProblemDetails = {
  type?: string
  title?: string
  status?: number
  detail?: string
  instance?: string
  [key: string]: unknown
}

export class ApiError extends Error {
  status: number
  problem?: ProblemDetails

  constructor(message: string, status: number, problem?: ProblemDetails) {
    super(message)
    this.name = 'ApiError'
    this.status = status
    this.problem = problem
  }
}

File: backend/app/src/main/kotlin/com/guardiangrow/app/http/ProblemDetails.kt
package com.guardiangrow.app.http

import kotlinx.serialization.Serializable

@Serializable
data class ProblemDetails(
  val type: String? = null,
  val title: String,
  val status: Int,
  val detail: String? = null,
  val instance: String? = null,
)

fun problemDetails(
  title: String,
  status: Int,
  detail: String? = null,
  instance: String? = null,
  type: String? = null,
) = ProblemDetails(type = type, title = title, status = status, detail = detail, instance = instance)
